user-manual/src/config/kdl.md

```diff
@@ -255,6 +264,121 @@ Filters at this stage are the earliest. Currently supported filters:
     * Arguments: `key="KEY" value="VALUE"`, where `KEY` is a valid HTTP header key, and `VALUE` is a valid HTTP header value
     * The given header will be added or replaced to `VALUE`
 
+### `services.$NAME.rate-limiting`
+
+This section contains the configuration for rate limiting rules.
+
+Rate limiting rules are used to limit the total number of requests made by downstream clients,
+based on various criteria.
+
+Note that Rate limiting is on a **per service** basis, services do not share rate limiting
+information.
+
+This section is optional.
+
+Example:
+
+```
+rate-limiting {
+    timeout millis=100
+
+    rule kind="source-ip" \
+        max-buckets=4000 tokens-per-bucket=10 refill-qty=1 refill-rate-ms=10
+
+    rule kind="uri" pattern="static/.*" \
+        max-buckets=2000 tokens-per-bucket=20 refill-qty=5 refill-rate-ms=1
+}
+```
+
+#### `services.$NAME.rate-limiting.timeout`
+
+The `timeout` parameter is used to set the total timeout for acquiring all rate limiting tokens.
+
+If acquiring applicable rate limiting tokens takes longer than this time, the request will not be
+forwarded and will respond with a 429 error.
+
+This parameter is mandatory if the `rate-limiting` section is present.
+
+This is specified in the form:
+
+`timeout millis=TIME`, where `TIME` is an unsigned integer
+
+**Implementation Note**: The rate limiting timeout is a temporary implementation detail to limit
```

Comment:
Naive implementation question: why have a timeout, or a max debt, at all -- why can't the unavailability of a token immediately cause a 429?

Reply:
That's a good question! The hope was to support features like "delay" used [by nginx](https://blog.nginx.org/blog/rate-limiting-nginx).

However, until [leaky-bucket has support for limiting 'max debt'](https://github.com/udoprog/leaky-bucket/issues/34), it might be better to treat tokens as "now or never" like you proposed. I had missed it, but the `leaky-bucket` crate [does support this](https://docs.rs/leaky-bucket/latest/leaky_bucket/struct.RateLimiter.html#method.try_acquire), so I think I'll remove the timeout parameter, and only offer "request succeeds or fails immediately" for now!